### PR TITLE
Updated comms model

### PR DESCRIPTION
**Pull request by caguero**
A pass of improvements over the communication model when breadcrumbs are deployed. 

The main change is a modification on how distance is computed in the model. Before this change, the Euclidean distance between source and destination was used even in the presence of breadcrumbs. This created a radial effect in the probability of message success as shown in the picture. This is inconsistent with the fact that breadcrumbs should extend the communication range. In the following example the source robot is in the staging area.

comms_v1

The new approach still takes into account distance but in a different way. Instead, we compute the greatest single hop of the route from source to destination. A hop is the distance from source to the first breadcrumb, between a pair of breadcrumbs, or between the last breadcrumb and the destination.  In the following example with the updated model the source robot is in the staging area.

comms_v2

The updated model also adds a small penalty (in distance) every time that a breadcrumb is added to the route.

This should close issue #488 .

### CODE REVIEW

**Comment by acschang** [comment]
**For tracking sake:**
Example breadcrumb placement in PW01 is as follows:
pw01_v2_test1

There are three issues from this breadcrumb placement with this PR:
pw01_option_c_issues

1. Comms dead zone in green area
2. Comms dead zone between drop-off in the Y tile and the adjacent tile
3. Low comms escalating to high comms in further adjacent tile

These issues are not present with the current comms model (A) or visualization-only comms model (B):
pw01_legacy_bc_exp

File: subt_ign/src/VisibilityTable.cc [hunk 1]
```diff
@@ -348,29 +378,109 @@ bool VisibilityTable::PopulateVisibilityInfoHelper(
...
+  const std::vector<ignition::math::graph::VertexId> &_relaySequence,
+  const ignition::math::graph::VertexId &_to) const
+{
+  if (_relaySequence.size() < 2)
+    return 0;
+
+  auto firstBreadcrumb = _relaySequence.at(0);
+  auto secondBreadcrumb = _relaySequence.at(1);
+
+  // We only consider the first breadcrumb stored in the tile.
```
Comment (acschang) [inline subt_ign/src/VisibilityTable.cc:467]: Is it accurate that additional breadcrumbs in a tile have no bearing on the position max distance single hop?

**Review by acschang — COMMENTED** [review]
Results look good on test cases with the resolution of all but the first issue (dead zone) which is not related to a source modification.  If you can clarify the comment I left on `VisibilityTable.cc` then I think we can be good to go on a merge here.

File: subt_ign/src/VisibilityTable.cc (hunk 1 above)
Comment (caguero) [reply subt_ign/src/VisibilityTable.cc:467]: > Is it accurate that additional breadcrumbs in a tile have no bearing on a max distance single hop from that tile?
> 
> e.g.
> Placing a breadcrumb at the beginning and end of a long tile would have an adverse effect on communication beyond the max hop distance would be unable to factor in the breadcrumb at the end of the long tile.

That's correct, we're not taking into account additional breadcrumbs within the same tile.

Comment (acschang) [reply subt_ign/src/VisibilityTable.cc:467]: This could be related to https://github.com/osrf/subt/issues/587

**Comment by angelacmaio** [comment]
Please update the [documentation](https://github.com/osrf/subt/wiki/Breadcrumbs-and-communication-visualization-tutorial) with more detail on how the comms model works once this is merged.

**Review by acschang — APPROVED** [review]
This is ready as issues 2 and 3 have been addressed.  A subsequent PR to address overlapping tiles blocking communication can be used to resolve issue 1 as that issue is not related to these changes and is present in the current communication model and the visibility-only communication model.

File: subt_ign/include/subt_ign/VisibilityTypes.hh [hunk 2]
```diff
@@ -20,10 +20,39 @@
...
 {
+
+/// \brief A class that contains some information about the cost of
+/// communicating two robots.
+class VisibilityCost
+{
+  /// \brief The cost from one tile to another tile.
+  public: double cost;
+
+  /// \brief The best route connecting source and destination. No that route
```
[the remark this answers is not in the document]
Comment (caguero) [reply subt_ign/include/subt_ign/VisibilityTypes.hh]: Thanks, adf9ec9

**Pull request merged**